### PR TITLE
Refactor gtest-parallel.

### PR DESCRIPTION
**Pull request by ehlemur-zz**
What I tried to do here is to create some classes to store the information about the tests and test executions so they're not saved in multiple times, and move things from the global scope to functions to improve readability.
I also introduced the capability to retry failed tests.

### CODE REVIEW

**Comment by ehlemur-zz** [comment]
@pbos @kwiberg 
Please take a look.

File: gtest-parallel [hunk 1]
```diff
@@ -30,6 +30,10 @@ import threading
 import time
 import zlib
 
+
+INF = float("inf")
```
Comment (pbos) [inline gtest-parallel:34]: Can you use None instead?

**Comment by pbos** [comment]
Is it possible for you to break this up into smaller changes? The linediff is huge, and you're making the file about 20% larger.

**Comment by ehlemur-zz** [comment]
I'll try.
If it's not too complex, can you take an overall view of the changes and tell me what you think?

**Pull request closed without merging**